pex/layout.py

```diff
@@ -126,18 +149,243 @@ def record(self, dest_dir):
+                    layout.extract_dist(
+                        dest_dir=spread_chroot.work_dir,
+                        dist_relpath=dist_relpath,
+                        is_wheel_file=deps_are_wheel_files,
+                    )
+
+        safe_mkdir(os.path.dirname(symlink_dest))
+        os.symlink(symlink_src, symlink_dest)
+
+
+def _ensure_distributions_installed_parallel(
+    layout,  # type: _Layout
+    pex_info,  # type: PexInfo
+    work_dir,  # type: str
+    install_to,  # type: str
+    max_jobs,  # type: int
+):
+    # type: (...) -> None
+
+    from textwrap import dedent
+
+    from pex.interpreter import spawn_python_job
+    from pex.jobs import SpawnedJob, execute_parallel
+
+    deps_are_wheel_files = pex_info.deps_are_wheel_files
+    install_cache = pex_info.install_cache
+
+    def install_distribution(item):
+        # type: (Tuple[str, str]) -> SpawnedJob[None]
+
+        location, sha = item
+        spread_dest = os.path.join(install_cache, sha, location)
+        dist_relpath = os.path.join(DEPS_DIR, location)
+        source = None if deps_are_wheel_files else layout.dist_strip_prefix(location)
+        extracting_message = (
+            "Installing wheel file {dist_relpath}".format(dist_relpath=dist_relpath)
+            if deps_are_wheel_files
+            else "Extracting {layout_type} distribution {dist_relpath}".format(
+                layout_type=layout.type, dist_relpath=dist_relpath
+            )
+        )
+        symlink_src = os.path.relpath(
+            spread_dest,
+            os.path.join(install_to, os.path.dirname(dist_relpath)),
+        )
+        symlink_dest = os.path.join(work_dir, dist_relpath)
+
+        return SpawnedJob.wait(
+            job=spawn_python_job(
+                args=[
+                    "-c",
+                    dedent(
+                        """\
+                        import os
+
+                        from pex.atomic_directory import atomic_directory
+                        from pex.common import safe_mkdir
+                        from pex.layout import identify_layout
+                        from pex.tracer import TRACER
+
+
+                        with identify_layout({pex!r}) as layout, atomic_directory(
+                            {spread_dest!r}, source={source!r}
+                        ) as spread_chroot:
+                            if not spread_chroot.is_finalized():
+                                with TRACER.timed({extracting_msg!r}):
+                                    layout.extract_dist(
+                                        dest_dir=spread_chroot.work_dir,
+                                        dist_relpath={dist_relpath!r},
+                                        is_wheel_file={is_wheel_file!r}
+                                    )
+
+                        symlink_dest = {symlink_dest!r}
+                        safe_mkdir(os.path.dirname(symlink_dest))
+                        os.symlink({symlink_src!r}, symlink_dest)
```

Comment:
Is it worth exposing the core of this logic as a function that can be used by both both `_ensure_distributions_installed_serial` directly and by this script?

Reply:
I'm not sure how you get much better. This code is pretty small and the sticking point is 7 parameters.